toolchain/README.md

```diff
@@ -0,0 +1,168 @@
+# Toolchain
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+The toolchain represents the production portion of Carbon. The toolchain's top
+priority is performance: it needs to generate code quickly and scale well.
+
+The main compiler is `//toolchain/driver:carbon`. When compiling, the current
+flow of data is:
+
+1. Load the file into a [SourceBuffer](source/source_buffer.h).
+2. Lex a SourceBuffer into a [TokenizedBuffer](lexer/tokenized_buffer.h).
+3. Parse a TokenizedBuffer into a [ParseTree](parser/parse_tree.h).
+4. Transform a ParseTree into a [SemanticsIR](semantics/semantics_ir.h).
+5. This flow is still incomplete: code generation, using LLVM, is still
+   required.
+
+## Lexing
+
+The [TokenizedBuffer](lexer/tokenized_buffer.h) is the central point of lexing.
+
+TODO: Not sure what's worth noting here.
+
+## Parsing
+
+The [ParseTree](parser/parse_tree.h) is the output of parsing, but most logic is
+in [ParserImpl](parser/parser_impl.h).
+
```

Comment:
```suggestion

The parse tree faithfully represents the tree structure of the source program, interpreted according to the Carbon grammar. No semantics are associated with the tree structure at this level, and no name lookup is performed.

Each parse tree node has an expected structure, corresponding to the grammar of the Carbon language, and the parser ensures that a valid parse tree node always has a valid structure. However, any parse tree node can be marked as invalid, and an invalid parse tree node can contain child nodes of any kind in any order. This is intended to model the situation where parsing failed because the code did not match the grammar, but we were still able to parse some subexpressions, as an aid for non-compiler tools such as syntax highlighters or refactoring tools.

Many functions in the parser return `llvm::Optional<T>`. A return value of `llvm::None` indicates that parsing has failed and an error diagnostic has already been produced, and that the current region of the parse tree might not meet its invariants so that the caller should create an invalid parse tree node. Other return values indicate that parsing was either successful or that any encountered errors have been recovered from, so the caller can create a valid parse tree node.

```